### PR TITLE
Feat/cache curated communities

### LINKED ISSUE #12277: Cache Community portal results in a table so that we don't have to wait for the contract call
### Description

Split up version of https://github.com/status-im/status-desktop/issues/7418

Start async process to fetch curated communities as soon as the user logs in. Then once this data is fetched for the first time it should be cached, probably in a table, so that the next time the user logs in, if they navigate to the community directory, and the latest data has not been fetched, then the cached data should be displayed instead.

One way to do it is to maybe have a field in the communities table about whether the community is curated and/or featured.
Then just have a query that retrieves only those.

### Acceptance Criteria

- Curated communities are cached some way or  another
- Opening the app and going to the curated communities portal shows the previous results even before the contract answers

### PR DESCRIPTION
**Pull request by osmaczko**
- fetch contract every two minutes and save results in database
- always return database entries to clients

closes: https://github.com/status-im/status-desktop/issues/12277

### CODE REVIEW

File: protocol/communities/persistence.go [hunk 1]
```diff
@@ -1632,3 +1632,70 @@ func (p *Persistence) SaveRequestsToJoin(requests []*RequestToJoin) (err error)
...
+		if featured {
+			result.ContractFeaturedCommunities = append(result.ContractFeaturedCommunities, communityID)
+		}
+	}
+
+	return result, nil
+}
+
+func (p *Persistence) SetCuratedCommunities(communities *CuratedCommunities) error {
+	tx, err := p.db.BeginTx(context.Background(), &sql.TxOptions{})
```
Comment (MishkaRogachev) [inline protocol/communities/persistence.go:1663]: i wish we can boilerplate transaction someday (not a comment, just sharing)

File: protocol/communities/persistence.go [hunk 2]
```diff
@@ -1632,3 +1632,70 @@ func (p *Persistence) SaveRequestsToJoin(requests []*RequestToJoin) (err error)
...
+	}
+	defer stmt.Close()
+
+	featuredMap := make(map[string]bool)
+	for _, community := range communities.ContractFeaturedCommunities {
+		featuredMap[community] = true
+	}
+
+	for _, community := range communities.ContractCommunities {
+		_, err := stmt.Exec(community, featuredMap[community])
```
Comment (MishkaRogachev) [inline protocol/communities/persistence.go:1694]: if we save all curated communities as featured, can it be just `_, err := stmt.Exec(community, true)` ?

**Review by MishkaRogachev — COMMENTED** [review]
Great work!  Just one question about saving curated communities..

File: protocol/messenger_curated_communities.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+				}
+
+				err = m.communitiesManager.SetCuratedCommunities(curatedCommunities)
+				if err == nil {
+					cache = curatedCommunities
+				} else {
+					logger.Error("failed to save curated communities", zap.Error(err))
+				}
+
+				response, err := m.fetchCuratedCommunities(curatedCommunities)
```
Comment (jrainville) [inline protocol/messenger_curated_communities.go:59]: Can we just pass `curatedCommunities` to `SendCuratedCommunitiesUpdate`. Maybe we can just wrap it in the Response?

**Review by jrainville — COMMENTED** [review]
Nice job. Just one small thing

File: protocol/communities/persistence.go (hunk 2 above)
Comment (osmaczko) [reply protocol/communities/persistence.go:1694]: We don't save all curated communities as featured, see how `featuredMap` is propagated, am I missing something :thinking: ?

File: protocol/communities/persistence.go (hunk 1 above)
Comment (osmaczko) [reply protocol/communities/persistence.go:1663]: Wdym exactly?

File: protocol/messenger_curated_communities.go (hunk 3 above)
Comment (osmaczko) [reply protocol/messenger_curated_communities.go:59]: Not sure I am following :thinking:  The response is a type crafted specifically for clients, including fetched community data, not only the IDs of curated and featured communities.

File: protocol/communities/persistence.go (hunk 1 above)
Comment (MishkaRogachev) [reply protocol/communities/persistence.go:1663]: we have this code repeated dozens of times. Just curious, if it's possible to boilerplate it 
```
tx, err := p.db.BeginTx(context.Background(), &sql.TxOptions{})

if err != nil {
		return err
	}

defer func() {
		if err == nil {
			err = tx.Commit()
			return
		}
		// don't shadow original error
		_ = tx.Rollback()
	}()
```

File: protocol/communities/persistence.go (hunk 2 above)
Comment (MishkaRogachev) [reply protocol/communities/persistence.go:1694]: Ah, i see, it's `communities.ContractFeaturedCommunities`

**Review by MishkaRogachev — APPROVED** [review]

File: protocol/communities/persistence.go (hunk 1 above)
Comment (osmaczko) [reply protocol/communities/persistence.go:1663]: Ah, would be nice indeed. The most obvious would be to use helper function:

```
defer func() {
    FinalizeTransaction(tx, &err)
}()
```

**Pull request merged**